dev/PowerNotifications/PowerNotifications.h

```diff
@@ -0,0 +1,709 @@
+            // as an integer percentage value from 0 to 100.
+            auto fullChargedCapacity = compositeBatteryStatus.Information.FullChargedCapacity;
+            auto remainingCapacity = compositeBatteryStatus.Status.Capacity;
+            if (fullChargedCapacity == BATTERY_UNKNOWN_CAPACITY ||
+                fullChargedCapacity == 0)
+            {
+                throw winrt::hresult_error(E_FAIL, L"Unknown charge ratio: FullChargedCapacity Unknown");
+            }
+            else if (remainingCapacity == BATTERY_UNKNOWN_CAPACITY)
+            {
+                throw winrt::hresult_error(E_FAIL, L"Unknown charge ratio: RemainingCapacity Unknown");
+            }
+            else if (remainingCapacity > fullChargedCapacity) {
+                std::wstring message = L"High charge ratio: RemainingCapacity " + std::to_wstring(remainingCapacity)
+                    + L" FullChargedCapacity " + std::to_wstring(fullChargedCapacity);
+                throw winrt::hresult_error(E_FAIL, message);
+
+            }
+            else if (remainingCapacity > (INT_MAX / 200))
+            {
+                std::wstring message = L"Absurd capacity:: RemainingCapacity " + std::to_wstring(remainingCapacity)
+                    + L" FullChargedCapacity " + std::to_wstring(fullChargedCapacity);
+                throw winrt::hresult_error(E_FAIL, message);
+            }
+            else
+            {
+                auto newRemainingChargePercent = static_cast<int>((remainingCapacity * 200) / fullChargedCapacity);
+                newRemainingChargePercent += 1;
+                newRemainingChargePercent /= 2;
+                m_batteryChargePercent = newRemainingChargePercent;
+            }
+
+            auto powerState = compositeBatteryStatus.Status.PowerState;
+
+            // Set battery status
+            if (compositeBatteryStatus.ActiveBatteryCount == 0)
+            {
+                m_batteryStatus = BatteryStatus::NotPresent;
+            }
+            else if (WI_IsFlagSet(powerState, BATTERY_DISCHARGING))
+            {
+                m_batteryStatus = BatteryStatus::Discharging;
+            }
+            else if (WI_IsFlagSet(powerState, BATTERY_CHARGING))
+            {
+                m_batteryStatus = BatteryStatus::Charging;
+            }
+            else
+            {
+                m_batteryStatus = BatteryStatus::Idle;
+            }
+
+            // Set power supply state
+            if (WI_IsFlagClear(powerState, BATTERY_POWER_ON_LINE))
+            {
+                m_powerSupplyStatus = PowerSupplyStatus::NotPresent;
+            }
+            else if (WI_IsFlagSet(powerState, BATTERY_DISCHARGING))
+            {
+                m_powerSupplyStatus = PowerSupplyStatus::Inadequate;
+            }
+            else
+            {
+                m_powerSupplyStatus = PowerSupplyStatus::Adequate;
+            }
+        }
+
+        void FireCorrespondingCompositeBatteryEvent()
```

Comment:
There is no locking here, so bad things happen if two callbacks occur at the same time (say because the batter status and energy saver status both changed at the same time).